### PR TITLE
Container infrastructure for Pulsar Reader

### PR DESCRIPTION
**Pull request by sobychacko**
Adding the basic foundations for the Pulsar Reader container. This is similar to the message listener container mechanism used in the regular consumer for managing the various lifecycles.

Resovles https://github.com/spring-projects-experimental/spring-pulsar/issues/341

### CODE REVIEW

File: spring-pulsar/src/test/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainerTests.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+				pulsarReaderFactory, readerContainerProperties);
+		container.start();
+
+		Map<String, Object> prodConfig = Map.of("topicName", "dprlct-001");
+		DefaultPulsarProducerFactory<String> pulsarProducerFactory = new DefaultPulsarProducerFactory<>(pulsarClient,
+				prodConfig);
+		PulsarTemplate<String> pulsarTemplate = new PulsarTemplate<>(pulsarProducerFactory);
+		pulsarTemplate.sendAsync("hello john doe");
+		assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
+		container.stop();
```
Comment (onobc) [inline spring-pulsar/src/test/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainerTests.java:88]: I think we should start this test out early on by always stopping the container in a finally block. I reworked this test to use `var` and `try-finally` and it looks like the following. I think it is still clean w/ the try/finally. 

```java
	@Test
	void latestMessageId() throws Exception {
		var latch = new CountDownLatch(2);
		var containerProps = new PulsarReaderContainerProperties();
		containerProps.setReaderListener((ReaderListener<?>) (reader, msg) -> {
			assertThat(msg.getValue()).isEqualTo("This message should be received by the reader");
			latch.countDown();
		});
		containerProps.setStartMessageId(MessageId.latest);
		containerProps.setTopics(List.of("dprlct-003"));
		containerProps.setSchema(Schema.STRING);

		var readerConfig = Collections.<String, Object>emptyMap();
		var readerFactory = new DefaultPulsarReaderFactory<String>(pulsarClient, readerConfig);
		DefaultPulsarReaderListenerContainer<String> container = null;
		try {
			container = new DefaultPulsarReaderListenerContainer<>(readerFactory, containerProps);
			
			var prodConfig = Map.<String, Object>of("topicName", "dprlct-003");
			var producerFactory = new DefaultPulsarProducerFactory<>(pulsarClient, prodConfig);
			var pulsarTemplate = new PulsarTemplate<>(producerFactory);

			// The following sends will not be received by the reader as we are using the
			// latest message id to start from.
			for (int i = 0; i < 5; i++) {
				pulsarTemplate.sendAsync("This message should not be received by the reader");
			}
			container.start();

			pulsarTemplate.sendAsync("This message should be received by the reader");
			pulsarTemplate.sendAsync("This message should be received by the reader");

			assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
		}
		finally {
			safeStopContainer(container);
		}
	}
```

**Review by onobc — CHANGES_REQUESTED** [review]
@sobychacko looking good! I know this is a 1st pass but I went ahead and captured my thoughts and requested some changes. Once you have adjusted what you are going to adjust in this particular pr go ahead and merge so you continue making forward progress. 

Thanks.

File: spring-pulsar/src/main/java/org/springframework/pulsar/core/DefaultPulsarReaderFactory.java [hunk 2]
```diff
@@ -54,7 +55,9 @@ public DefaultPulsarReaderFactory(PulsarClient pulsarClient, Map<String, Object>
 	public Reader<T> createReader(@Nullable List<String> topics, @Nullable MessageId messageId, Schema<T> schema)
 			throws PulsarClientException {
 		ReaderBuilder<T> readerBuilder = this.pulsarClient.newReader(schema);
-		readerBuilder.topics(topics);
+		if (!CollectionUtils.isEmpty(topics)) {
+			readerBuilder.topics(topics);
+		}
 		readerBuilder.startMessageId(messageId);
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/core/DefaultPulsarReaderFactory.java:61]: It does and the reader has a default of starting from the `latest`.

File: spring-pulsar/src/main/java/org/springframework/pulsar/event/ReaderFailedToStartEvent.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+public class ReaderFailedToStartEvent extends PulsarEvent {
+
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * Construct an instance with the provided source and container.
+	 * @param source the container instance that generated the event.
+	 * @param container the container or the parent container if the container is a child.
+	 */
+	public ReaderFailedToStartEvent(Object source, Object container) {
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/event/ReaderFailedToStartEvent.java:28]: Issue created: https://github.com/spring-projects-experimental/spring-pulsar/issues/343

File: spring-pulsar/src/main/java/org/springframework/pulsar/reader/AbstractPulsarReaderListenerContainer.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	@Override
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/reader/AbstractPulsarReaderListenerContainer.java:133]: https://github.com/spring-projects-experimental/spring-pulsar/issues/344

File: spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+import org.springframework.core.task.AsyncTaskExecutor;
+import org.springframework.core.task.SimpleAsyncTaskExecutor;
+import org.springframework.pulsar.core.PulsarReaderFactory;
+import org.springframework.pulsar.event.ReaderFailedToStartEvent;
+import org.springframework.pulsar.event.ReaderStartedEvent;
+import org.springframework.pulsar.event.ReaderStartingEvent;
+import org.springframework.scheduling.SchedulingAwareRunnable;
+
+/**
+ * Default implementation for the Pulsar reader container.
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java:40]: https://github.com/spring-projects-experimental/spring-pulsar/issues/345

File: spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ * @param <T> reader data type.
+ * @author Soby Chacko
+ */
+public class DefaultPulsarReaderListenerContainer<T> extends AbstractPulsarReaderListenerContainer<T> {
+
+	private final AtomicReference<InternalAsyncReader> internalAsyncReader = new AtomicReference<>();
+
+	private volatile CountDownLatch startLatch = new CountDownLatch(1);
+
+	private volatile CompletableFuture<?> readerFuture;
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java:54]: https://github.com/spring-projects-experimental/spring-pulsar/issues/346

File: spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+		@Override
+		public void run() {
+			DefaultPulsarReaderListenerContainer.this.readerThread.set(Thread.currentThread());
+			publishReaderStartingEvent();
+			publishReaderStartedEvent();
+
+			while (isRunning()) {
+				try {
+					Message<T> message = this.reader.readNext();
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/reader/DefaultPulsarReaderListenerContainer.java:162]: https://github.com/spring-projects-experimental/spring-pulsar/issues/347

File: spring-pulsar/src/main/java/org/springframework/pulsar/reader/PulsarReaderContainerProperties.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+import org.springframework.core.task.AsyncTaskExecutor;
+import org.springframework.util.Assert;
+
+/**
+ * Container properties for Pulsar {@link org.apache.pulsar.client.api.Reader}.
+ *
+ * @author Soby Chacko
+ */
+public class PulsarReaderContainerProperties {
```
[the remark this answers is not in the document]
Comment (sobychacko) [reply spring-pulsar/src/main/java/org/springframework/pulsar/reader/PulsarReaderContainerProperties.java:33]: https://github.com/spring-projects-experimental/spring-pulsar/issues/348

**Pull request merged**